### PR TITLE
 [NUI] Load Tizen.NUI dependencies libs on Windows and MacOS

### PR DESCRIPTION
**Pull request by Coquinho**
\### Description of Change ###
This update `.csproj` to automatically define an `MSBuild` compiler
directive `NOTIZEN` when building the project on a MacOS or Windows
systems.
This directive selects which dll names to use when compiling, thus
enabling `Tizen.NUI` to correctly search for libs at Mac and Windows.

This impacts the load of:
- `dali-csharp-binder`
- `capi-appfw-app-manager`
- `capi-appfw-app-common`
- `dlog`

We first tried to define those names at run-time, but
`DllImportAtribbuts` are `C#` `const strings`, which need to be defined
at compile-time, so it raised the need for the preprocessor directive to
select the correct name.

\### API Changes ###
There are no API changes, but as BSD `readlink` and `expr` are different than
the one on Linux, this PR workaround those functions at `build.sh`, creating
platform-independent functions that run at BSD and Linux `Bash` as well as 
Windows `Git Bash`, while maintaining the desired output on both systems.

### CODE REVIEW

**Comment by WonyoungChoi** [comment]
@Coquinho It is dangerous to automatically decide `NOTIZEN` constant at the build time because many TizenFX developers build their modules for Tizen in Windows and MacOS environment using IDE such like VS and VSCode. I think that it would be better to manually pass a special property to build.sh and process the property in build/common.props for changing the DllImport behaviour.
for example
```
./build.sh /p:UseOnlyDllName=true
```

File: build.sh [hunk 1]
```diff
@@ -170,7 +190,7 @@ cmd_pack() {
     dotnet pack --no-restore --no-build --nologo -o $OUTDIR \
            $SCRIPT_DIR/build/pack.csproj \
            /p:Version=$VERSION \
-           /p:NuspecFile=$(readlink -f $nuspec)
+           /p:NuspecFile=$(realpath $nuspec)
```
Comment (WonyoungChoi) [inline build.sh:193]: Need `dash` here also. 
```bash
  -p:Version=$VERSION \
  -p:NuspecFile=$(realpath $nuspec)
```

**Comment by WonyoungChoi** [comment]
@Coquinho readlink and expr in my Git Bash of Windows are working well. my `coreutils` version is 8.32. 
```bash
$ expr "8.0.0.12345" : '\([0-9]\+\.[0-9]\+\.[0-9]\+\)'
8.0.0
$ readlink -f .
/d/Work/TizenFX

$ expr --version
expr (GNU coreutils) 8.32
Copyright (C) 2020 Free Software Foundation, Inc.
....
```
Could you let me know the version of yours?

**Comment by Coquinho** [comment]
> @Coquinho It is dangerous to automatically decide `NOTIZEN` constant at the build time because many TizenFX developers build their modules for Tizen in Windows and MacOS environment using IDE such like VS and VSCode. I think that it would be better to manually pass a special property to build.sh and define constants with the property in build/common.props for changing the DllImport behaviour.
> for example
> 
> ```
> ./build.sh full -p:NOTIZEN=true
> ```

Oh sure, didn't know about that, I will work on it. 
On Windows I am able to pass the definition almost like what you said, as in #2091 :
```
$ ./build.sh full -p:DefineConstants=NOTIZEN
```
But for some reason, in MacOS, it wasn't being passed as expected, and the wrong dll name was being used.
I will give it a new shot and see if I can make it work in MacOS.

File: build.sh (hunk 1 above)
Comment (Coquinho) [reply build.sh:193]: Oh, sorry about that, thanks for noticing

**Comment by Coquinho** [comment]
Ok, so I've tried some other approaches:

- Reverting the `.csproj` changes and passing `-p:DefineConstants=NOTIZEN` at command line:
    ```bash
    $ ./build.sh full -p:DefineConstants=NOTIZEN
    ```
    As I am testing this at NUI.Sample, I've done:
    ```bash
    $ dotnet build test/Tizen.NUI.Samples/Tizen.NUI.Samples/Tizen.NUI.Samples.csproj /property:GenerateFullPaths=true /consoleloggerparameters:NoSummary -p:DefineConstants=NOTIZEN
    $ dotnet run test/Tizen.NUI.Samples/Tizen.NUI.Samples/bin/Debug/netcoreapp3.1/Tizen.NUI.Samples.dll --project test/Tizen.NUI.Samples/Tizen.NUI.Samples/Tizen.NUI.Samples.csproj /property:GenerateFullPaths=true /consoleloggerparameters:NoSummary -p:DefineConstants=NOTIZEN
    ```
- Reverting the `.csproj`changes letting this instead:
    ```c#
    <PropertyGroup Condition="'$(NOTIZEN)'=='true'">
        <DefineConstants>$(DefineConstants);NOTIZEN</DefineConstants>
    </PropertyGroup>
    ```
    And passing `-p:NOTIZEN=true` as you suggested:
    ```bash
    $ ./build.sh full -p:NOTIZEN=true
    $ dotnet build test/Tizen.NUI.Samples/Tizen.NUI.Samples/Tizen.NUI.Samples.csproj /property:GenerateFullPaths=true /consoleloggerparameters:NoSummary -p:NOTIZEN=true
    $ dotnet run test/Tizen.NUI.Samples/Tizen.NUI.Samples/bin/Debug/netcoreapp3.1/Tizen.NUI.Samples.dll --project test/Tizen.NUI.Samples/Tizen.NUI.Samples/Tizen.NUI.Samples.csproj /property:GenerateFullPaths=true /consoleloggerparameters:NoSummary -p:NOTIZEN=true
    ```
None of those worked in MacOS: when running NUI Sample it tried to load the wrong lib name at both. The first one I've tested, and I know it works on Windows.

- Reverting the `.csproj`changes, letting the `NOTIZEN`condition as earlier:
    ```c#
    <PropertyGroup Condition="'$(NOTIZEN)'=='true'">
        <DefineConstants>$(DefineConstants);NOTIZEN</DefineConstants>
    </PropertyGroup>
    ```
    But now setting `NOTIZEN=true` environment variable:
    ```bash
    $ ./build.sh full
    $ dotnet build test/Tizen.NUI.Samples/Tizen.NUI.Samples/Tizen.NUI.Samples.csproj /property:GenerateFullPaths=true /consoleloggerparameters:NoSummary
    $ NOTIZEN=true dotnet run test/Tizen.NUI.Samples/Tizen.NUI.Samples/bin/Debug/netcoreapp3.1/Tizen.NUI.Samples.dll --project test/Tizen.NUI.Samples/Tizen.NUI.Samples/Tizen.NUI.Samples.csproj /property:GenerateFullPaths=true /consoleloggerparameters:NoSummary
    ```
    Do work on MacOS.
    And on Windows running at vccode after a  `setx NOTIZEN "true"` worked, as well as passing `-p:DefineConstants:NOTIZEN` as in #2091 
What do you think, is this a good approach?

File: src/Tizen.Applications.Common/Tizen.Applications.Common.csproj [hunk 2]
```diff
@@ -4,6 +4,10 @@
     <TargetFramework>netstandard2.0</TargetFramework>
   </PropertyGroup>
 
+  <PropertyGroup Condition="'$(NOTIZEN)'=='true'">
+    <DefineConstants>$(DefineConstants);NOTIZEN</DefineConstants>
+  </PropertyGroup>
+
```
Comment (WonyoungChoi) [inline src/Tizen.Applications.Common/Tizen.Applications.Common.csproj:10]: These lines can be moved to `build/common.props`
 the `build/common.props` is processed before each module's csproj.

**Comment by WonyoungChoi** [comment]
@Coquinho Using property `NOTIZEN` (current your changes) is better than using `DefineConstants` directly. :+1:

File: src/Tizen.Applications.Common/Tizen.Applications.Common.csproj (hunk 2 above)
Comment (Coquinho) [reply src/Tizen.Applications.Common/Tizen.Applications.Common.csproj]: Much better, thanks a lot for the review.

**Comment by Coquinho** [comment]
As GitHub was asking for a rebase, I've took the opportunity to squash the PR to its original structure.

**Comment by dongsug-song** [comment]
@gotoextreme 
Hello, 
This patch modifies Tizen.Application.Common part. 
could you review this?
thank you.

**Review by hyunjushin — DISMISSED** [review]
Please re-push this patch into DevelNUI branch. DevelNUI will be integrated with master branch on every week.

**Comment by dongsug-song** [comment]
@Coquinho 
Hello,
as far as I know, TizenFX is based on tizen c# platform and mostly supports tizen platform and tizen device.
I am not sure but I don't think only by this patch, NUI would run on Windows and Ubuntu, because DALi Ubuntu backend or DALi Windows backend is not included in this TizenFX package.
so I think it would be dangerous to add this kind of platform dependent patch in tizen platform.
please consider closing this patch here and instead, please check the following repositories to add your modification.
thank you.

- NUI Windows backend : https://github.sec.samsung.net/NUI/windows-backend
- NUI Ubuntu backend : https://code.sec.samsung.net/confluence/display/GFX/NUI+running+on+Ubuntu+backend

**Comment by walac** [comment]
Hi @dongsug-song ,

Here is Wander, @Coquinho 's co-worker. We have been working on porting DALi and NUI to WIndows and macOS. afaik, there is a team in Samsung working on shipping all dependencies necessary in the form a SDK. Our work is to enable that be possible.

**Comment by dongsug-song** [comment]
@gotoextreme 
Hello, this patch changes AppFW's codes.
could you review and give any comment if there is something wrong?
thank you.

**Comment by dongsug-song** [comment]
@gotoextreme
Hello, this patch changes AppFW's codes.
could you review and give any comment if there is something wrong?
thank you.

@Coquinho 
Hello, 
could you resolve the conflicts?
thank you.

**Comment by Coquinho** [comment]
> @Coquinho
> Hello,
> could you resolve the conflicts?
> thank you.

Sure, I will look at it, thanks.

**Comment by gotoextreme** [comment]
> @gotoextreme
> Hello, this patch changes AppFW's codes.
> could you review and give any comment if there is something wrong?
> thank you.
> 
> @Coquinho
> Hello,
> could you resolve the conflicts?
> thank you.

AppFW codes modifications looks good to me.

**Review by dongsug-song — APPROVED** [review]
looks good to me

**Comment by dongsug-song** [comment]
@Coquinho 
Hello, 
I can't merge your PR. error is shown and "Rebase and merge" button is inactivated.
could you check?

image

**Comment by Coquinho** [comment]
> @Coquinho
> Hello,
> I can't merge your PR. error is shown and "Rebase and merge" button is inactivated.
> could you check?
> 
> image


I've rebased the branch, so it should be fine now.

**Comment by dongsug-song** [comment]
@Coquinho 
Hello, 
still "Rebase and merge" button is inactivated. could you try to do merge in your side?
sorry but I have no idea how to make this button activated.

**Comment by Coquinho** [comment]
Hi @dongsug-song,

I have no access to merge it:
image

**Comment by Coquinho** [comment]
Hey @dongsug-song, I had a look at Github configs and repos can be set to only allow to merge updated branches. What I think is happening is that between me updating and you trying to merge it, some patch is merged, thus mine is no more updated, and you are not allowed to merge it. I will try to let this updated when I stop using my computer so that you can have a chance to merge it.

**Comment by dongsug-song** [comment]
@Coquinho
Hello,
it is same. not solved. still I can not merge because "Rebase and merge" button has been inactivated with the error message of "This branch is out-of-date with the base branch".
therefore, please close this PR and upload new same PR again.
thank you.

**Comment by dongsug-song** [comment]
@Coquinho 
Hello, 
please resolve the conflicts. otherwise please upload new PR and close this PR.
thank you.

**Comment by dongsug-song** [comment]
@Coquinho
Hello,
This PR has been pending for a long time, so could you resolve the conflicts by next week (03/07)
otherwise, I will close this PR without any notice.
thank you.

**Comment by dongsug-song** [comment]
@Coquinho
Hello
now I put the \*\***NO MERGE**\*\* tag on thie PR.
you can remove \*\***NO MERGE**\*\* tag for yourself when this PR is fixed and ready to be merged.
thank you.

**Pull request closed without merging**